druid/src/menu/mod.rs

```diff
@@ -0,0 +1,790 @@
+    }
+}
+
+/// An entry in a menu.
+///
+/// An entry is either a [`MenuItem`], a submenu (i.e. [`Menu`]), or one of a few other
+/// possibilities (such as one of the two options above, wrapped in a [`MenuLensWrap`]).
+pub struct MenuEntry<T> {
+    inner: Box<dyn MenuVisitor<T>>,
+}
+
+type MenuPredicate<T> = Box<dyn FnMut(&T, &T, &Env) -> bool>;
+
+/// A menu.
+///
+/// Menus can be nested arbitrarily, so this could also be a submenu.
+/// See the [module level documentation](crate::menu) for more on how to use menus.
+pub struct Menu<T> {
+    rebuild_on: Option<MenuPredicate<T>>,
+    refresh_on: Option<MenuPredicate<T>>,
+    item: MenuItem<T>,
+    children: Vec<MenuEntry<T>>,
+    // bloom?
+}
+
+impl<T: Data> From<Menu<T>> for MenuEntry<T> {
+    fn from(menu: Menu<T>) -> MenuEntry<T> {
+        MenuEntry {
+            inner: Box::new(menu),
+        }
+    }
+}
+
+type MenuCallback<T> = Box<dyn FnMut(&mut MenuEventCtx, &mut T, &Env)>;
+type HotKeyCallback<T> = Box<dyn FnMut(&T, &Env) -> Option<HotKey>>;
+
+/// An item in a menu.
+///
+/// See the [module level documentation](crate::menu) for more on how to use menus.
+pub struct MenuItem<T> {
+    id: MenuItemId,
+
+    title: LabelText<T>,
+    callback: Option<MenuCallback<T>>,
+    hotkey: Option<HotKeyCallback<T>>,
+    selected: Option<Box<dyn FnMut(&T, &Env) -> bool>>,
+    enabled: Option<Box<dyn FnMut(&T, &Env) -> bool>>,
+
+    // The last resolved state of this menu item. This is basically consists of all the properties
+    // above, but "static" versions of them not depending on the data.
+    old_state: Option<MenuItemState>,
+}
+
+impl<T: Data> From<MenuItem<T>> for MenuEntry<T> {
+    fn from(i: MenuItem<T>) -> MenuEntry<T> {
+        MenuEntry { inner: Box::new(i) }
+    }
+}
+
+struct Separator;
+
+impl<T: Data> From<Separator> for MenuEntry<T> {
+    fn from(s: Separator) -> MenuEntry<T> {
+        MenuEntry { inner: Box::new(s) }
+    }
+}
+
+impl<T: Data> Menu<T> {
+    /// Create an empty menu.
+    pub fn empty() -> Menu<T> {
+        Menu {
+            rebuild_on: None,
+            refresh_on: None,
+            item: MenuItem::new(""),
+            children: Vec::new(),
+        }
+    }
+
+    /// Create a menu with the given name.
+    pub fn new(title: impl Into<LabelText<T>>) -> Menu<T> {
+        Menu {
+            rebuild_on: None,
+            refresh_on: None,
+            item: MenuItem::new(title),
+            children: Vec::new(),
+        }
+    }
+
+    /// Provide a callback for determining whether this item should be enabled.
+    ///
+    /// Whenever the callback returns `true`, the menu will be enabled.
+    pub fn enabled(mut self, enabled: impl FnMut(&T, &Env) -> bool + 'static) -> Self {
+        self.item = self.item.enabled(enabled);
+        self
+    }
+
+    /// Append a menu entry to this menu, returning the modified menu.
+    pub fn with_entry(mut self, entry: impl Into<MenuEntry<T>>) -> Self {
```

Comment:
It doesn't *really* matter, but I might consider including the old method names as aliases, like:

```rust
#[doc(hidden)]
#[deprecated(since = "0.8.0", note = "use with_entry instead")]
pub fn append(mut self, entry: impl Into<MenuEntry<T>>) -> Self {
    self.with_entry(entry)
}
```

I might also consider having,

```rust
#[deprecated(since = "0.8.0", note = "Renamed to Menu")]
pub type MenuDesc<T> = Menu<T>;
```

anyway I'm sort of anal-retentive about this sort of stuff, the breakage isn't a big deal.